gatling-jms/src/main/scala/io/gatling/jms/action/JmsRequestSendAction.scala

```diff
@@ -0,0 +1,117 @@
+import io.gatling.core.stats.StatsEngine
+import io.gatling.core.util.NameGen
+import io.gatling.jms.client.JmsClient
+import io.gatling.jms.protocol.JmsProtocol
+import io.gatling.jms.request._
+
+/**
+ * Core JMS Action to handle Send
+ *
+ * This handles the core "send"ing of messages. Gatling calls the execute method to trigger a send.
+ */
+class JmsRequestSendAction(attributes: JmsAttributes, protocol: JmsProtocol, val statsEngine: StatsEngine, val next: Action)
+    extends ExitableAction with NameGen {
+
+  override val name = genName("jmsSend")
+
+  // Create a client to refer to
+  val client = JmsClient(protocol, attributes.destination)
+
+  /**
+   * Framework calls the execute() method to send a single request
+   * <p>
+   * Note this does not catch any exceptions (even JMSException) as generally these indicate a
+   * configuration failure that is unlikely to be addressed by retrying with another message
+   */
+  override def execute(session: Session): Unit = recover(session) {
+
+    val messageProperties = resolveProperties(attributes.messageProperties, session)
+
+    val jmsType = resolveOptionalExpression(attributes.jmsType, session)
+
+    // sendtime
+    val startDate = nowMillis
+
+    val msg = jmsType.flatMap(jmsType =>
+      messageProperties.flatMap { props =>
+        attributes.message match {
+          case BytesJmsMessage(bytes) => bytes(session).map(bytes => client.sendBytesMessage(bytes, props, jmsType))
+          case MapJmsMessage(map)     => map(session).map(map => client.sendMapMessage(map, props, jmsType))
+          case ObjectJmsMessage(o)    => o(session).map(o => client.sendObjectMessage(o, props, jmsType))
+          case TextJmsMessage(txt)    => txt(session).map(txt => client.sendTextMessage(txt, props, jmsType))
+        }
+      })
+
+    // done time
+    val endDate = nowMillis
+
+    msg.map { msg =>
+      executeNext(session, startDate, endDate, OK, next, attributes.requestName)
+    }
+
+  }
+
+  private def executeNext(
+    session:  Session,
+    sent:     Long,
+    received: Long,
+    status:   Status,
+    next:     Action,
+    title:    String,
+    message:  Option[String] = None
+  ) = {
+
+    val timings = ResponseTimings(sent, received)
+    statsEngine.logResponse(session, title, timings, status, None, message)
+    next ! session.logGroupRequest(timings.responseTime, status).increaseDrift(nowMillis - received)
+  }
+
+  def resolveProperties(
+    properties: Map[Expression[String], Expression[Any]],
+    session:    Session
+  ): Validation[Map[String, Any]] = {
+    properties.foldLeft(Map.empty[String, Any].success) {
+      case (resolvedProperties, (key, value)) =>
+        for {
+          key <- key(session)
+          value <- value(session)
+          resolvedProperties <- resolvedProperties
+        } yield resolvedProperties + (key -> value)
+    }
+  }
+
+  def logMessage(text: String, msg: Message): Unit = {
+    if (logger.underlying.isDebugEnabled()) {
```

Comment:
No, that doesn't change anything. `text` has already been computed (typical usage is String interpolation/concat), so debug ebaled test has to be done at caller site.